lib/src/hex_util.rs

```diff
@@ -12,29 +12,79 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#![allow(missing_docs)]
+//! Hex string helpers.
 
 const REVERSE_HEX_CHARS: &[u8; 16] = b"zyxwvutsrqponmlk";
 
-fn to_forward_hex_digit(b: u8) -> Option<u8> {
-    let value = match b {
-        b'k'..=b'z' => b'z' - b,
-        b'K'..=b'Z' => b'Z' - b,
-        _ => return None,
-    };
-    if value < 10 {
-        Some(b'0' + value)
-    } else {
-        Some(b'a' + value - 10)
+fn reverse_hex_value(b: u8) -> Option<u8> {
+    match b {
+        b'k'..=b'z' => Some(b'z' - b),
+        b'K'..=b'Z' => Some(b'Z' - b),
+        _ => None,
     }
 }
 
-pub fn to_forward_hex(reverse_hex: impl AsRef<[u8]>) -> Option<String> {
-    reverse_hex
-        .as_ref()
-        .iter()
-        .map(|b| to_forward_hex_digit(*b).map(char::from))
-        .collect()
+/// Decoded hex or error, which may include odd-length byte.
+#[derive(Clone, Debug)]
+pub enum DecodeHexResult {
+    /// Bytes decoded from even-length hex string.
+    Ok(Vec<u8>),
+    /// Bytes decoded from odd-length hex string, with the padded last byte.
+    Odd(Vec<u8>),
+    /// Hex string had invalid character.
+    InvalidChar,
+}
+
+impl DecodeHexResult {
+    fn from_decoded(decoded: Option<(Vec<u8>, bool)>) -> Self {
+        match decoded {
+            Some((bytes, false)) => DecodeHexResult::Ok(bytes),
+            Some((bytes, true)) => DecodeHexResult::Odd(bytes),
+            None => DecodeHexResult::InvalidChar,
+        }
+    }
+
+    /// Extracts bytes decoded from even-length hex string.
+    pub fn ok(self) -> Option<Vec<u8>> {
```

Comment:
Since some callers consider odd-length inputs okay, should we call this `.even()` or `.even_ok()` or similar?

We would perhaps need to rename `.ok_or_odd()` then (since we're no longer saying that "even == okay"), but that's also related to my other comment.

Reply:
I consider that odd-length hex is usually an error, hence `Ok|Odd`.

Should we add `decode_reverse_hex() -> Option<_>` and `decode_reverse_hex_maybe_odd() -> Option<(_, bool)>`?

Reply:
Yes, I think I would like that better, thanks. Then we won't need `DecodeHexResult` at all, right?

Reply:
Added `decode_reverse_hex()` and `decode_reverse_hex_prefix()`.